core/modules/wcontrol/Foreman.cc

```diff
@@ -39,353 +38,56 @@
-               std::shared_ptr<wbase::SendChannel> replyChannel) {
-
-        wbase::Task::Ptr t = std::make_shared<wbase::Task>(taskMsg, replyChannel);
-        _foremanImpl.newTaskAction(t);
-        std::shared_ptr<Cancel> c
-            = std::make_shared<Cancel>(t, _foremanImpl._scheduler);
-        return c;
-    }
-    ForemanImpl& _foremanImpl;
-};
-
-////////////////////////////////////////////////////////////////////////
-// ForemanImpl
-////////////////////////////////////////////////////////////////////////
-ForemanImpl::ForemanImpl(Scheduler::Ptr s)
-    : _scheduler(s), _log(LOG_GET("Foreman")),
-      _running(std::make_shared<wbase::TaskQueue>()) {
+Foreman::Foreman(Scheduler::Ptr s) : _scheduler{s} {
     // Make the chunk resource mgr
     mysql::MySqlConfig c(wconfig::getConfig().getSqlConfig());
     _chunkResourceMgr = wdb::ChunkResourceMgr::newMgr(c);
-
     _rManager.reset(new RunnerMgr(*this));
     assert(s); // Cannot operate without scheduler.
-
 }
-ForemanImpl::~ForemanImpl() {
+Foreman::~Foreman() {
+	LOGF(_log, LOG_LVL_DEBUG, "Foreman::~Foreman()");
     // FIXME: Poison and drain runners.
     // This should only (get called on shutdown/restart).
 }
 
-void
-ForemanImpl::_startRunner(wbase::Task::Ptr t) {
-    std::thread thrd{Runner{*_rManager, t}};
-    thrd.detach();
+/// Create and queue a Task from a TaskMsg and a replyChannel.
+wbase::Task::Ptr Foreman::processMsg(std::shared_ptr<proto::TaskMsg> const& taskMsg,
+                                     std::shared_ptr<wbase::SendChannel> const& replyChannel) {
+    wbase::Task::Ptr t = std::make_shared<wbase::Task>(taskMsg, replyChannel);
+    newTaskAction(t);
+    return t;
 }
 
-bool ForemanImpl::squashByHash(std::string const& hash) {
-    // Most (all?) cancellations are invoked by client with the
-    // active/outstanding request. This should only get called when a client not
-    // directly holding the active query wishes to cancel the query.
-    std::lock_guard<std::mutex> m(_mutex);
-    bool success = _scheduler->removeByHash(hash);
-    success = success || _rManager->squashByHash(hash);
-    return success;
+void Foreman::_startRunner(wbase::Task::Ptr const& t) {
+    auto f = [this](wbase::Task::Ptr const& t) {
+        std::unique_ptr<Runner> rp{new Runner{*(this->_rManager), t}};
+        (*rp)();
+    };
+    std::thread thrd{f, t};
+    thrd.detach();
 }
 
-void ForemanImpl::newTaskAction(wbase::Task::Ptr task) {
+void Foreman::newTaskAction(wbase::Task::Ptr const& task) {
     // Pass to scheduler.
     assert(_scheduler);
-    wbase::TaskQueuePtr newReady;
-    {
-        std::lock_guard<std::mutex> lock(_runnersMutex);
-        newReady = _scheduler->newTaskAct(task, _running);
-    }
+    auto newReady = _rManager->queueTask(task, _scheduler);
     // Perform only what the scheduler requests.
     if(newReady.get() && (newReady->size() > 0)) {
         wbase::TaskQueue::iterator i = newReady->begin();
```

Comment:
I expect it will change significantly with the next ticket which deals with queueing on the worker.